google/cloud/spanner/sql_statement.cc

```diff
@@ -0,0 +1,40 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/sql_statement.h"
+#include <algorithm>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+bool operator==(SqlStatement const& lhs, SqlStatement const& rhs) {
+  std::string s1 = lhs.statement_;
+  std::string s2 = rhs.statement_;
+  std::transform(std::begin(s1), std::end(s1), std::begin(s1),
+                 [](unsigned char c) { return std::tolower(c); });
+  std::transform(std::begin(s2), std::end(s2), std::begin(s2),
```

Comment:
It's too bad that we need to copy both statement strings. What do you think about writing a little helper function (in an unnamed namespace in this file) that takes two `std::string const&` arguments and does a case insensitive comparison without copying? It could quickly look at string sizes to short circuit the loop, then iterate both strings comparing the results of `std::tolower(s1[i]) != std::tolower(s2[i])`... or something?

Reply:
Done.